clients/da-vinci-client/src/main/java/com/linkedin/davinci/blob/BlobSnapshotManager.java

```diff
@@ -0,0 +1,124 @@
+    this.snapshotRetentionTime = snapshotRetentionTime;
+    this.snapShotTimestamps = new HashMap<>();
+    this.concurrentSnapshotUsers = new HashMap<>();
+  }
+
+  /**
+   * Checks if the current snapshot of the partition is stale
+   */
+  private boolean isSnapshotStale(String topicName, int partitionId) {
+    if (!snapShotTimestamps.containsKey(topicName) || !snapShotTimestamps.get(topicName).containsKey(partitionId)) {
+      return true;
+    }
+    return System.currentTimeMillis() - snapShotTimestamps.get(topicName).get(partitionId) > snapshotRetentionTime;
+  }
+
+  /**
+   * Updates the snapshot of the hybrid store
+   */
+  private void updateHybridSnapshot(RocksDB rocksDB, String topicName, int partitionId) {
+    String fullPathForPartitionDBSnapshot = RocksDBUtils.composeSnapshotDir(this.basePath, topicName, partitionId);
+    if (fullPathForPartitionDBSnapshot.isEmpty()) {
+      LOGGER.error(
+          "Error creating snapshot directory for topic: {} and partition: {} due to not path existing",
+          topicName,
+          partitionId);
+      return;
+    }
+    try {
+      Checkpoint checkpoint = createCheckpoint(rocksDB);
+
+      LOGGER.info("Creating snapshots in directory: {}", fullPathForPartitionDBSnapshot);
+      checkpoint.createCheckpoint(fullPathForPartitionDBSnapshot);
+
+      LOGGER.info("Finished creating snapshots in directory: {}", fullPathForPartitionDBSnapshot);
+    } catch (RocksDBException e) {
+      throw new VeniceException(
+          "Received exception during RocksDB's snapshot creation in directory " + fullPathForPartitionDBSnapshot,
+          e);
+    }
+  }
+
+  /**
+   * Checks if the snapshot is stale, if it is and no one is using it, it updates the snapshot,
+   * otherwise it increases the count of people using the snapshot
+   */
+  public void getHybridSnapshot(RocksDB rocksDB, String topicName, int partitionId) {
+    if (rocksDB == null || topicName == null) {
+      throw new IllegalArgumentException("RocksDB instance and topicName cannot be null");
+    }
+    String fullPathForPartitionDBSnapshot = RocksDBUtils.composeSnapshotDir(this.basePath, topicName, partitionId);
+    if (!snapShotTimestamps.containsKey(topicName)) {
+      snapShotTimestamps.put(topicName, new HashMap<>());
+      concurrentSnapshotUsers.put(topicName, new HashMap<>());
+    }
+    if (isSnapshotStale(topicName, partitionId)) {
+      if (concurrentSnapshotUsers.get(topicName).get(partitionId) == null
+          || concurrentSnapshotUsers.get(topicName).get(partitionId).get() == 0) {
+        updateHybridSnapshot(rocksDB, topicName, partitionId);
+        snapShotTimestamps.get(topicName).put(partitionId, System.currentTimeMillis());
+      }
+      if (getConcurrentUsers(rocksDB, topicName, partitionId) == 0) {
+        concurrentSnapshotUsers.get(topicName).put(partitionId, new AtomicLong(1));
+      } else {
+        concurrentSnapshotUsers.get(topicName).get(partitionId).incrementAndGet();
```

Comment:
Also, this method does not seem to be thread-safe. But, I'd suggest we address other comments first, and eventually we come back to add the safeguard for thread safety.

Reply:
I think given the fact that we believe snapshot operations are fast, we can use a single lock object to guard against the 2 operations: (1) Check and maybe increment the snapshot user count (2) Decrease the snapshot user count.

One example in the codebase is: DiskHealthCheckService. Can you check if you can follow the usage there to use a single lock to guard against the operations?


After you are done, can you write a unit test about this behavior?
In order to test it properly, feel free to add setter/getter method with default modifier to set/get the snapshot map for your testing purpose.